core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java

```diff
@@ -111,17 +112,19 @@ protected void processTasks(CancelTasksRequest request, Consumer<CancellableTask
     }
 
     @Override
-    protected synchronized TaskInfo taskOperation(CancelTasksRequest request, CancellableTask cancellableTask) {
+    protected synchronized void taskOperation(CancelTasksRequest request, CancellableTask cancellableTask,
+            ActionListener<TaskInfo> listener) {
         final BanLock banLock = new BanLock(nodes -> removeBanOnNodes(cancellableTask, nodes));
         Set<String> childNodes = taskManager.cancel(cancellableTask, request.getReason(), banLock::onTaskFinished);
         if (childNodes != null) {
             if (childNodes.isEmpty()) {
                 logger.trace("cancelling task {} with no children", cancellableTask.getId());
-                return cancellableTask.taskInfo(clusterService.localNode(), false);
+                listener.onResponse(cancellableTask.taskInfo(clusterService.localNode().getId(), false));
             } else {
                 logger.trace("cancelling task {} with children on nodes [{}]", cancellableTask.getId(), childNodes);
+                // TODO wait for this to complete before firing the listener?
```

Comment:
I don't think we have to wait. This is the listener that is responsible for notifying the user that their operation to cancel the task is accepted. There is another listener in BanLock that is getting triggered when the ban is acknowledged by the node AND the task finishes.